src/Features/Core/Portable/Diagnostics/BuildOnlyDiagnosticsService.cs

```diff
@@ -6,29 +6,48 @@
 using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Collections;
 using Microsoft.CodeAnalysis.Host;
 using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Diagnostics;
 
 [ExportWorkspaceServiceFactory(typeof(IBuildOnlyDiagnosticsService), ServiceLayer.Default), Shared]
 [method: ImportingConstructor]
 [method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
-internal sealed class BuildOnlyDiagnosticsServiceFactory() : IWorkspaceServiceFactory
+internal sealed class BuildOnlyDiagnosticsServiceFactory(
+    IAsynchronousOperationListenerProvider asynchronousOperationProvider) : IWorkspaceServiceFactory
 {
     public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
-        => new BuildOnlyDiagnosticsService(workspaceServices.Workspace);
+        => new BuildOnlyDiagnosticsService(workspaceServices.Workspace, asynchronousOperationProvider.GetListener(FeatureAttribute.Workspace));
 
-    private sealed class BuildOnlyDiagnosticsService : IBuildOnlyDiagnosticsService
+    private sealed class BuildOnlyDiagnosticsService : IBuildOnlyDiagnosticsService, IDisposable
     {
-        private readonly object _gate = new();
+        private readonly CancellationTokenSource _disposalTokenSource = new();
+        private readonly AsyncBatchingWorkQueue<Func<CancellationToken, Task>> _workQueue;
```

Comment:
💡 It would probably be better to use an object to represent the work instead of a callback.

Reply:
The other option is to just make it a queue of `WorkspaceChangeEventArgs`.

Reply:
yup.  nice suggestion.  updated to that.